tutorial: Transfer Files

haadcode opened the pull request:
This PR will add a basic interop example as per https://github.com/ipfs/js-ipfs/issues/711.

diasdavid commented on examples/interop-examples/README.md:
```diff
@@ -0,0 +1,102 @@
+# interop examples
```
`s/interop examples/fetch content added by go-ipfs`

diasdavid commented on examples/interop-examples/README.md:
````diff
@@ -0,0 +1,102 @@
+npm start
+```
+
+Open http://127.0.0.1:8080 in your favorite browser.
````
Chrome or Firefox

haadcode replied:
Both should work, thus "your favorite" :)

diasdavid replied:
There is also Safari, Edge, Internet Explorer and Opera, not everyone's favourite is Chrome or Firefox :)

diasdavid commented on examples/interop-examples/README.md:
```diff
@@ -0,0 +1,102 @@
+Steps
+1. create IPFS instance
+
+First, we need an IPFS instance that we'll use throughout the tutorial. For the sake of convinience, we'll use `ipfs-daemon` module to create and IPFS instance.
```
We should not be recommending `ipfs-daemon` in order to instantiate `js-ipfs`, it just means that `js-ipfs` needs to improve its init process.

haadcode replied:
I'll rework this in a way that the init dance is in its own file and we include `js-ipfs` instead of `ipfs-daemon`. 

I don't think we'll get to merge ipfs-daemon to js-ipfs within two weeks given how massive public API change that is. Let me know if you think we can.

diasdavid replied:
> I don't think we'll get to merge ipfs-daemon to js-ipfs within two weeks given how massive public API change that is. Let me know if you think we can.

We totally can, it is not actually that big of a change, really, just requires putting in the hours, it is pretty clear what we want out of it and we have a ton of tests creating js-ipfs daemons with different configurations and in Node.js + Browser, those that will ensure that the new API works.

It is, however, a low priority compared to the rest of the tasks for this Sprint, so, if we get to finish all the others, we will want ot tackle this.

Until then, the demo should be without an external module to create an instance of this module.

diasdavid commented on examples/interop-examples/README.md:
````diff
@@ -0,0 +1,102 @@
+})
+```
+
+2. listen for file drop
````
The goal of this demo is to "access files from go-ipfs", and not adding files inside the browser.

haadcode replied:
Understood.

Somewhere we discussed, and decided, that we do go-ipfs -> js-ipfs-browser -> js-ipfs-browser (stream a video or smth). I believe adding the support for adding files to the browser version is not that much out of scope in terms of work and it adds another important step to the tutorial for one of the most asked questions for someone who's starting with ipfs "how do I add files to IPFS?". So having that functionality gives a more fully fledged tutorial than just catting files from go-ipfs. Don't worry, we're still going for the same goal: access files from go-ipfs.

diasdavid replied:
Thank you :)

haadcode commented:
> The demo available here is not an interop demo, in fact, it doesn't touch go-ipfs at all, I know you wanted first to do js-ipfs to js-ipfs, but that is not the purpose.

All of this is what the example will be. Perhaps the intent was dropped in the communication. If you take a look at my sprint updates, you'll see the progress being described (ie. it's WIP).

See this comment https://github.com/ipfs/js-ipfs/pull/714#discussion_r96379582 for details to why there's browser-to-browser functionality too.

haadcode commented:
To clarify, this PR will (eventually) implement the example as designed here: https://github.com/ipfs/js-ipfs/issues/711#issuecomment-272805392

diasdavid commented:
@haadcode thank you for clarifying! :)

diasdavid commented on examples/interop-examples/README.md:
````diff
@@ -0,0 +1,102 @@
+
+## Run
+```
+cd examples/interop-examples
````
no need, people will be inside this folder already

diasdavid commented on examples/interop-examples/package.json:
```diff
@@ -0,0 +1,18 @@
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "ipfs-daemon": "^0.3.0-beta.16"
```
When `ipfs-daemon` is out of the equation, it will be so much faster to get this example to work because one less npm install of the whole thing is needed.

diasdavid commented on examples/interop-examples/package.json:
```diff
@@ -0,0 +1,18 @@
+    "start": "http-server -c-1"
+  },
+  "author": "",
+  "license": "ISC",
```
s/ISC/MIT

diasdavid commented:
@haadcode I was going to review this example again, but I see there hasn't been any other commit since my last review, so I will avoid listing out the points of feedback again.

I received a notification through IRC that you were preparing some handoff guide for the next person to take on this examples, however, I'm being unable to find it, I've scanned the IRC logs and opened all the related github issues, could you point me to it? Thank you.

haadcode commented:
> you were preparing some handoff guide for the next person to take on this examples

Working on it.

haadcode commented:
Updated everything, fixed a bunch of stuff. This should be now ready to be picked up by others. What remains to be done is listed in: https://github.com/ipfs/js-ipfs/blob/e916a5d1d0c654a6111e5a4861aafb79b72b3e62/examples/access-go-ipfs-files/cat-a-file/README.md#todo. There's also the new signal server address that needs to be fixed as per https://github.com/libp2p/js-libp2p-webrtc-star/issues/87.

The short is: add the UI for connecting to a go-ipfs address manually and make sure catting the file from go-ipfs-->js-ipfs(-browser) works.

diasdavid commented:
Seems that it is assumed that the user has ran `npm run build` before hand.

```
» npm i

> cat-a-file@1.0.0 postinstall /Users/koruza/code/js-ipfs/examples/access-go-ipfs-files/cat-a-file
> cp ../../../dist/index.js ./public/ipfs.js

cp: ../../../dist/index.js: No such file or directory

npm WARN cat-a-file@1.0.0 No repository field.
npm ERR! Darwin 16.1.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "i"
npm ERR! node v6.9.1

```

diasdavid commented:
The example doesn't have a way to signal a want to dial to any other node, assumes that `webrtc` is used. The purpose was to offer to the user a way to dial through websockets.

diasdavid commented:
working on this example today

diasdavid commented:
## Todo notes left by @haadcode for @RichardLitt 

- Add "connect to peer" input field and "connect" button under the "Peers" section in index.html
- Add `connectToPeer` function which calls `ipfs.swarm.connect(multiaddr)`. See https://github.com/ipfs/js-ipfs/blob/b0a7cd83cbf146b0f147467dedc686f94a5f751f/examples/ipfm/src/DataStore.js#L82 and https://github.com/ipfs/js-ipfs/blob/b0a7cd83cbf146b0f147467dedc686f94a5f751f/examples/ipfm/README.md#start-an-interplanetary-file-exchange-daemon. The multiaddr to connect to looks like this: `/ip4/127.0.0.1/tcp/9999/ws/ipfs/QmZGH8GeASSkSZoNLPEBu1MqtzLTERNUEwh9yTHLEF5kcW`
- Hook up "connect" button's click event to `connectToPeer` function
- Add instructions to this README on how to add a file in go-ipfs
- Add instructions to this README on how to cat it in the UI
- Make sure the "Start a go-ipfs daemon" instructions are correct
- Make sure go-ipfs daemon and the example connect to each other

VictorBjelkholm commented:
Taking this over for now

haadcode commented:
Thanks @VictorBjelkholm for jumping on this and improving it!

Quick comment on the commits:
- It's better to use dynamic sizes (for fonts, margins, widths, etc.) instead of pixels, ie. `px` --> `em` / `rem` / etc. 
- When changing CSS, make sure to test the changes at least with Firefox and Chrome to make sure they both work (usually they don't quite render the same), and make sure to try it on a mobile browser(s) too to make sure the layouts and fonts scale. 
- I believe that you're still working on this, but make sure the picture is actually shown: https://github.com/ipfs/js-ipfs/pull/714/commits/ea48e5bb4288d6ed45358edc206529c71949c831#diff-6db58e31a051cfc7a871d0fd2444b037R100 :)

Good job so far! 👍

diasdavid commented:
@VictorBjelkholm just your latest changes, here is my feedback:

- The blue'ish theme definitely gives it more of a tone of a 'finished' thing
- The information about the js-ipfs node disappeared completely, it needs to come back
- The new guided focus makes it that the user is less in power, losing the capability of exchanging content between browsers
- It is forcing the user to dial to a 'go-ipfs' node, remember, this demo works with any node outside of the browser
- The default values have tiny boxes, making it hard to copy paste the values and make sure we are not doing something wrong

diasdavid commented:
The user should be more of in control, here is a sketch of what is intended

![img_0001](https://cloud.githubusercontent.com/assets/1211152/22378102/4c3b7b56-e4ac-11e6-8e6b-3a6fda4e5d07.JPG)

Comparing to the previous version, the only thing that was missing was really the `Connect` box

diasdavid commented:
![image](https://cloud.githubusercontent.com/assets/1211152/22387702/d191f276-e4d4-11e6-86be-61623937703e.png)


Let's avoid showing the signal server add option and always default to our hosted one, the reality is that everyone will always use the hosted and by putting that as an option in the front page, it will make people focus too much on 'the signalling thing'

diasdavid commented:
![image](https://cloud.githubusercontent.com/assets/1211152/22387731/035de440-e4d5-11e6-953c-8f1e0d5b14bb.png)

Let's take a bit more of the screen.

When I paste a multiaddr to dial, I would like to be able to see the whole thing that I am pasting.

diasdavid commented:
Also note, that the browser node should not have a TCP addr, remove it from the config.

VictorBjelkholm commented:
@diasdavid 
> Let's avoid showing the signal server add option and always default to our hosted one, the reality is that everyone will always use the hosted and by putting that as an option in the front page, it will make people focus too much on 'the signalling thing'

Thinking about it, why give the option of the data-dir even? We could just jump directly into the state where we're starting the daemon, or just showing the start daemon button. 

> Let's take a bit more of the screen.
> When I paste a multiaddr to dial, I would like to be able to see the whole thing that I am pasting.

Makes sense, didn't get to the peer section yet. Thanks!

> Also note, that the browser node should not have a TCP addr, remove it from the config.

Oh, I think it ships with that? The only thing we're doing now in this example is to add the webrtc-star multiaddr. You're saying that the TCP addr is only wrong in this example or in general of js-ipfs?

diasdavid commented:
> Thinking about it, why give the option of the data-dir even? We could just jump directly into the state where we're starting the daemon, or just showing the start daemon button.

Let's do that, skip that too, make it always random :) 

> Oh, I think it ships with that? The only thing we're doing now in this example is to add the webrtc-star multiaddr. You're saying that the TCP addr is only wrong in this example or in general of js-ipfs?

Yes, just checked again, it shouldn't (it doesn't cause any harm because it is filtered out, but it will be confusing for users)

VictorBjelkholm commented:
Pushed latest changes that makes this complete-ish

**Things missing:**
* [ ] replace .cat with .get (don't get any data in streams with .get for some reason, @diasdavid ideas?), so right now the downloading only support text files...
* [ ] Refactor code + add valuable comments
* [ ] Filter out `/ip4/0.0.0.0/tcp` address
* [ ] Fix drag-and-drop not closing popup when not dropping file in UI

diasdavid commented:
Can we figure out how to avoid having addrs in multiple lines?

![image](https://cloud.githubusercontent.com/assets/1211152/22398239/600aa5aa-e57c-11e6-9944-7c8a15044431.png)


Also s/Remote Peers/Connected Peers

diasdavid commented:
I see peers, but it is not fetching for me, it also doesn't print any error

VictorBjelkholm commented:
> Can we figure out how to avoid having addrs in multiple lines?

Yeah, either we have a scroll or the font-size has to be very small. Not sure

diasdavid commented:
@VictorBjelkholm here is your fix for the `filter TCP` item on your checklist https://github.com/ipfs/js-ipfs/pull/734

VictorBjelkholm commented:
Sorry @haadcode missed your comments before, reply inline:

> It's better to use dynamic sizes (for fonts, margins, widths, etc.) instead of pixels, ie. px --> em / rem / etc.
I would like the layout to stay the same no matter the viewport so I'll go with px here.

> When changing CSS, make sure to test the changes at least with Firefox and Chrome to make sure they both work (usually they don't quite render the same), and make sure to try it on a mobile browser(s) too to make sure the layouts and fonts scale.

Sure, no problem. I'm not gonna spend any time on trying it on mobile though, if people want it to work on mobile, they'll have to solve any issues themselves (or we can do another example for mobile in the future)

> I believe that you're still working on this, but make sure the picture is actually shown: ea48e5b#diff-6db58e31a051cfc7a871d0fd2444b037R100 :)

Not really, I've removed the picture since I didn't see how it tied into the example. I think it's not necessary to show a picture inline in the webpage but rather when people add files, we show them in a download list.

VictorBjelkholm commented:
### Last update before weekend

Still missing some refactor, mainly splitting up the ipfs and UI logic, right now it's all wrapped together. Made just one quick pass for comments so probably there are things that can be better explained. I have not looked at the tutorial and I'm pretty sure it's unfinished. Did some improvements with the drag-and-drop but still there is a bug left.

Biggest one is that `.get` doesn't actually work at the moment. Purpose is to fetch the content of the file and set as an octet-stream as the download when clicking the link, but something is crashing with the wrong cid version. @diasdavid is working on fixing this.

* [ ] Refactor code + add valuable comments
* [ ] Have .get actually work with just one file and also maybe directories
* [ ] Fix drag-and-drop not closing popup when ending up not dropping file in UI
* [ ] Finish tutorial in readme.md

diasdavid commented:
@VictorBjelkholm I just fresh git cloned and fresh install both package.json and this is what I'm getting:

![image](https://cloud.githubusercontent.com/assets/1211152/22418223/fc5eabb4-e6cf-11e6-8a43-97c503e4c604.png)

VictorBjelkholm commented:
Final pass-through before leaving this move on to the CI/CD sprint

* cd into /complete directory of the transfer files example

* Run `npm install`

* Run `npm start`

* `cd ../../../`

* run `npm run build`

* `cd examples/transfer-files/complete`

* `npm start`

* navigate to `localhost:12345`

* in js-ipfs, run `IPFS_PATH=$HOME/.examples-test node src/cli/bin.js init`

* run `IPFS_PATH=$HOME/.examples-test node src/cli/bin.js daemon`

* run `IPFS_PATH=$HOME/.examples-test node src/cli/bin.js id`

* grab the address containing `/ws`

* connect directly to it by putting `/ip4/127.0.0.1/tcp/4003/ws/ipfs/QmPs14AG8igTwHKvZ2deVDrkYTtczghHTeyivUNsSydhpH` in the multiaddr input and pressing "Connect to peer"

* Address shows up in list of remote peers

* Run `IPFS_PATH=$HOME/.examples-test node src/cli/bin.js add package.json`, save the hash

* Paste the hash into multihash input field, press "Fetch file"

* Date - Hash and size appears in the list of fetched files

* Clicking on the item for the file triggers download

* File finished downloading, opening up shows content of file being "content of file" since the `.get` is not fully working yet... See TODO comment in function `catFile` in `app.js`

* Dragging a file into the UI shows wonky "drop file popup" that appears and disappears when moving the mouse around

* Dropping the file shows "Added antigen.sh as Qm[...]" and changes the hash in the multihash input

* Clicking "Fetch file" adds the file to the list of files with it's size and the date

* Clicking on the added file item downloads the file but still with the file contents being "content of file"

* Let's try with a go-ipfs node. Kill the current running nodejs node

* Peer disappears from the UI in the complete example

* Pull down latest master of go-ipfs and install it

* Run `IPFS_PATH=$HOME/.examples-test-go ipfs init`

* Run `IPFS_PATH=$HOME/.examples-test-go ipfs daemon`

* No `/ws` address, need to add it to the configuration

* Open `$HOME/.examples-test-go/config` and add `/ip4/127.0.0.1/tcp/4004/ws` to the list of swarm addresses

* Run `IPFS_PATH=$HOME/.examples-test-go ipfs daemon`

* Run `IPFS_PATH=$HOME/.examples-test-go ipfs id`, save the `/ws` address

* Paste `/ws` address into multiaddr input and press "Connect to peer"

* Peer address shows up in UI

* Run `IPFS_PATH=$HOME/.examples-test-go ipfs add appveyor.yml` in go-ipfs repository

* Save hash

* Paste hash into multihash input and press "Fetch file"

* Receive stream with files but it doesn't have any `.content` field as the  response from nodejs process has

VictorBjelkholm commented:
Discovered via @diasdavid doing a demo today with this.

* [ ] Peers list have a comma between every peer that is not supposed to be there
* [ ] Multiaddr input is too wide and goes outside the containing box

diasdavid commented:
@VictorBjelkholm fixed the get problem. The stream was never resumed :) 

https://github.com/ipfs/js-ipfs/pull/714/commits/843e227637c4d5783b33a33e96ad2efa5ad9fe77

diasdavid commented:
Follow here: https://github.com/ipfs/js-ipfs/pull/774#issuecomment-284711991
